lib/code_corps_web/plugs/data_to_attributes.ex

```diff
@@ -12,13 +12,30 @@ defmodule CodeCorpsWeb.Plug.DataToAttributes do
   def init(opts), do: opts
 
   @spec call(Conn.t, Keyword.t) :: Plug.Conn.t
-  def call(%Conn{params: %{"data" => data} = params} = conn, _opts) do
+  def call(%Conn{params: %{"data" => data} = params} = conn, many_mappings) do
     attributes =
       params
       |> Map.delete("data")
-      |> Map.merge(data |> JaSerializer.Params.to_attributes)
+      |> Map.delete("included")
+      |> Map.merge(params |> parse_data)
+      |> Map.merge(params |> parse_included(many_mappings))
 
     conn |> Map.put(:params, attributes)
   end
   def call(%Conn{} = conn, _opts), do: conn
+
+  defp parse_data(%{"data" => data}), do: data |> JaSerializer.Params.to_attributes
+  defp parse_data(%{}), do: %{}
+
+  defp parse_included(%{"included" => included}, many_mappings) do
+    included |> Enum.reduce(%{}, fn (%{"data" => %{"type" => type}} = params, parsed) ->
+      attributes = params |> parse_data()
+
+      case many_mappings |> Map.has_key?(type) do
+        true -> parsed |> Map.update(many_mappings[type], [attributes], &List.push(&1, attributes))
+        false -> parsed |> Map.put(type, attributes)
+      end
+    end)
+  end
```

Comment:
The critical part of the `included` support is here.

I approached this by following the existing doc for this plug, where it's stated that the goal of the plug is to convert JSON API params into a map of params suitable for `Ecto.Changeset`

From that, the goal was

- Included `belongs_to` records are merged with params as a submap property, with the key identical to the belongs to record type
- Included `has_many` records are merged with params as a list property, with the key being a pluralized version of the record type

This allows us to simply call `cast_assoc(changeset, :record_type, opts)` and the casting works out of the box.

I did not work on inferring the pluralized type directly and instead opted to explicitly provide these when calling the plug. It should be sufficient for our needs, but there is a way to infer instead.

Reply:
Since the plug has become quite complex, I think we should add tests for it.